Cleanup; Add `main`. Change CLI parser to use Python stdlib's `argparse`. Configuration file.

tildebyte opened the pull request:
- PEP8!
- Move most end-of-line comments above their related code.
- Copyedit most comments.
- Add `if __name__ = __main__:` so that we're a proper command line program with a single point of entry and (mostly) single point of exit.
- Add nice command line argument parsing/args help using Python stdlib's `argparse` module

I came up with this patch independently of @Tursaanpoika's nice work in #8. I think a proper `main` is always appropriate for a stand-alone command line script.
@donaldm changes in #7 amount to a fork, IMHO. Also, I'm not sure why it's a good idea to use JSON (complex) over Python stdlib's `ConfigParser` (simple), or use the (almost) deprecated and more complicated `optparse` over `argparse`.

I'm looking next at pulling out the model vars (`model_path`, `net_fn`, `param_fn`) into a ConfigParser file, as I agree with @donaldm that it'd be nice for them to be easily configurable, and that it's a bit too involved for the command line. Also, it might be nice to option-ize the output image format and mention in the `--input` help some of the other formats which PIL can handle.

tildebyte committed:
Cleanup; Add `main` and change command line parser to use Python stdlib's `argparse`.
   - PEP8!
   - Move most end-of-line comments above their related code.
   - Copyedit most comments.
   - Add `if __name__ = __main__:` so that we're a proper command line program with a single point of entry and (mostly) single point of exit.
   - Add nice command line argument parsing/args help using Python stdlib's `argparse` module

I came up with this patch independently of @Tursaanpoika's issue 8 (https://github.com/Dhar/image-dreamer/issues/8). I think a proper `main` is always appropriate for a stand-alone command line script.
@donaldm's changes in issue 7 (https://github.com/Dhar/image-dreamer/issues/7) amount to a fork, IMHO. Also, I'm not sure why it's a good idea to use JSON over Python stdlib's much easier `ConfigParser`, or use the (almost) deprecated `optparse` rather than `argparse`.

tildebyte committed:
Fix program description; shebang; stub in version info for command line.

Tursaanpoika commented:
I think @donaldm https://github.com/Dhar/image-dreamer/issues/7 had one more option in the code, scale. I'll grab your source and tinker around a bit :D argparse is what google came up when I searched on how to add cli -options to python, I'm no developer, but I still do like to nose around to see what I can do :)

Tursaanpoika commented:
Oh, it was step scale, defaulting to 1.5

tildebyte committed:
Cleanup
   - A little more PEP8.
   - Remove unused imports.
   - Remove unneeded iPython notebook cruft.
   - `print` as a function.
   - Since we now have defaults coming from `argparse`, we no longer need to (or want to) have defaults hard-coded as named parameters to functions scattered throughout the code. Everything has been pulled up into the `deepdream` call from inside `main`. I don't know if it's practical or desirable to tweak the remaining parms (clip, step_size, jitter), but maybe they can be pulled out into a (future) config file.

tildebyte commented:
Right, I just spotted that myself and restructured things a bit.

donaldm commented:
I opted to use JSON because I could represent the image mean list could be directly represented in the JSON instead of having to parse it out. Also I think JSON is a nice format which can support additions to the format. 

I used OptionParser because I usually have to work with/support Python 2.4 (lol). I am just more familiar with it and I really don't see what argparse adds that I needed for this (but argparse works just as well :)).

I do agree my version is more of a fork though as I restructured things a good bit from the original code.

tildebyte commented:
Awesome. Thanks for the comments. I have a `ConfigParser` commit almost ready. For this usage, it's really simple; it's also really easy to extend. I  _really_ like the readability of the cfg file itself. 

I agree that parsing complex data could be a pain; I ran into the exact issue with `mean` which you mentioned ;)

donaldm commented:
Are you using ConfigParser or SafeConfigParser?

Also I think the concept of mean needs to be looked at, because there are mean files potentially in the models themselves that are not being used currently. May need to evaluate this.

tildebyte commented:
Safe.

tildebyte committed:
Pull internal parms out into a `ConfigParser` file.

tildebyte committed:
Continuing to: add stuff to config; abstract; get rid of magic variables where possible.

dreamify.py:
   - Make temp file name configurable.
   - Add temp file name to config file.
   - Process `mean` as part of configuration.
   - Pull some of the code blobs in `main` out into fns of their own, parameterizing and abstracting.
   - Add some comments based on code from @donaldm (https://github.com/Dhar/image-dreamer/issues/7) designed to help with using other models.

dreamify.cfg:
   - tempfile name.
   - Sample parm set to use Google's "places" dataset.

tildebyte committed:
dreamify.py: Pull `model_path` in from config file. Could be useful to reference e.g. mean data.
dreamify.cfg: Add more models.
